### DISCUSSION TITLE
How to fix erroneous occlusion rendering in 3D scenes?

### DISCUSSION THREAD

**Discussion by maria-korosteleva** [zauberzeug/nicegui#3514] [Q&A] [answered]
### Question

Hi! I'm trying to display a 3D scene consisting of a body model and a garment. There is a relatively small gap between the two, which seems to be confusing occulision estimation and creating weird images as below: 
![image](https://github.com/user-attachments/assets/0eff9e2d-76b0-422c-8f2f-73c6c3195dc3)

The same two 3D objects are rendered fine using an external renderer: 
![Configured_design_3D_render_front](https://github.com/user-attachments/assets/97bf8c2a-6aad-4fbb-9583-9293b4c794af)

My best guess is there should be a setting that influences z-buffer resolution, which influences the differentiation of the objects that are close to each other in the depth direction from the camera. However, I do not seem to find it in niceGUI. Is it available? 

Here is the loading code snippet, for reference: 
```py
with ui.scene(
	width=1024,
	height=800, 
	camera=camera, 
	grid=False, 
	background_color=bg_color   
	) as ui_3d_scene:
	ui_3d_scene.stl(
			'/body/mean_all.stl' 
		).rotate(np.pi / 2, 0., 0.).material(color='#000000')

	ui_3d_scene.gltf(
		f'geo/{self.garm_3d_filename}', 
	).scale(0.01).rotate(np.pi / 2, 0., 0.).material(side='double')

```

**Comment by rodja** [comment]
Can you render the model as expected with https://threejs.org/? Thats the framework we use for `ui.scene`.

**Reply by maria-korosteleva** [reply]
I can render in threejs editor without a problem (https://threejs.org/editor/)

![image](https://github.com/user-attachments/assets/856f30ca-2fdc-4ee0-b636-97e02c66f866)

**Comment by falkoschindler** [comment]
I'm trying to find a reproduction with simple geometric objects. Maybe like this:
```py
with ui.scene(width=800, height=600) as scene:
    scene.sphere().material(color='red').move(z=1.000)
    scene.sphere().material(color='blue').move(z=1.001)
```
![Screenshot 2024-08-16 at 08 49 37](https://github.com/user-attachments/assets/a2308ba6-4e06-42e1-8fbe-d71ae1ecff4a)

Strangely, I can't reproduce the artefact with plain Three.js:
```html
<html>
  <body>
    <script src="https://cdn.jsdelivr.net/npm/three@0.157.0/build/three.min.js"></script>
    <script>
      const scene = new THREE.Scene();
      scene.add(new THREE.Mesh(new THREE.SphereGeometry(1, 32, 16), new THREE.MeshBasicMaterial({ color: "red" })));
      scene.add(new THREE.Mesh(new THREE.SphereGeometry(1, 32, 16), new THREE.MeshBasicMaterial({ color: "blue" })));
      scene.children[0].position.y += 1.0;
      scene.children[1].position.y += 1.001;

      const camera = new THREE.PerspectiveCamera(75, 1.33, 0.1, 1000);
      camera.up = new THREE.Vector3(0, 1, 0);
      camera.position.set(0, 3, 8);
      camera.lookAt(0, 0, 0);

      const renderer = new THREE.WebGLRenderer({ antialias: true, alpha: true });
      renderer.setSize(800, 600);
      renderer.render(scene, camera);
      document.body.appendChild(renderer.domElement);
    </script>
  </body>
</html>
```
![Screenshot 2024-08-16 at 08 49 25](https://github.com/user-attachments/assets/a6fe91ee-c39f-45a5-90ca-daf4d9a187c6)

Does anyone have an idea what is going on?

**Reply by python-and-novella** [reply]
For me, it seems to be related with kernel version.
![QQ图片20240816211203](https://github.com/user-attachments/assets/abaf6425-81cc-41e6-9fd2-b3d09983ab8c)

**Reply by falkoschindler** [reply]
That's interesting! 😲

**Reply by maria-korosteleva** [reply]
I use Chrome Version 127.0.6533.120 (Official Build) (64-bit) 🤔 
The same models render fine in threejs editor but with artifacts in niceGUI ui.scene, on the same browser

**Reply by falkoschindler** [reply]
@python-and-fiction For me the behavior is the same for Chrome, Firefox and Safari (macOS 14.6.1).
Are you sure it is not only the camera angle causing different results? What if you add `scene.move_camera(10, 10)`. The farther away the worse the artifacts, as I noticed.

**Reply by python-and-novella** [reply]
@falkoschindler I do as you said, it likes this:
![image](https://github.com/user-attachments/assets/339c4888-29d2-45ae-b746-8a36648b8e2d)
Also, I tested the pure html.
```html
<html>
  <body>
    <script src="https://unpkg.com/three@0.157.0/build/three.js"></script>
    <script>
      const scene = new THREE.Scene();
      scene.add(new THREE.Mesh(new THREE.SphereGeometry(1, 32, 16), new THREE.MeshBasicMaterial({ color: "red" })));
      scene.add(new THREE.Mesh(new THREE.SphereGeometry(1, 32, 16), new THREE.MeshBasicMaterial({ color: "blue" })));
      scene.children[0].position.y += 1.0;
      scene.children[1].position.y += 1.001;

      const camera = new THREE.PerspectiveCamera(75, 1.33, 0.1, 1000);
      camera.up = new THREE.Vector3(0, 1, 0);
      camera.position.set(0, 3, 8);
      camera.lookAt(0, 0, 0);

      const renderer = new THREE.WebGLRenderer({ antialias: true, alpha: true });
      renderer.setSize(800, 600);
      renderer.render(scene, camera);
      document.body.appendChild(renderer.domElement);
    </script>
  </body>
</html>
```
It shows this:
![image](https://github.com/user-attachments/assets/ccdb7287-7cc9-4a2d-abea-2be777ed193d)
Maybe my chromium based browser do some magic in kernel, it performes unexpected, plz ignore it.
As for origin bug, you're right, the camera angle causing different results, problem will dispear when i zoom in.

**Comment by maria-korosteleva** [comment]
Added as an issue: https://github.com/zauberzeug/nicegui/issues/3673

**Comment by falkoschindler** [comment]
@maria-korosteleva Can you, please, share the STL files as well as a minimal executable code snippet so that we can reproduce the original problem? Thanks!

**Reply by maria-korosteleva** [reply]
[render_errors_example.zip](https://github.com/user-attachments/files/16896342/render_errors_example.zip)

Right, sorry, here ⬆️

The script is included as .py file, and here are the contents: 

```py
from nicegui import ui, app
import numpy as np

def create_camera(cam_location, fov):
    camera = ui.scene.perspective_camera(fov=fov)
    camera.x = cam_location[0]
    camera.y = cam_location[1] 
    camera.z = cam_location[2]

    # direction
    camera.look_at_x = 0
    camera.look_at_y = 0
    camera.look_at_z = cam_location[2] * 2/3

    return camera

def create_lights(scene:ui.scene, intensity=30.0):
        light_positions = np.array([
            [1.60614, 1.23701, 1.5341,],
            [1.31844, -2.52238, 1.92831],
            [-2.80522, 2.34624, 1.2594],
            [0.160261, 3.52215, 1.81789],
            [-2.65752, -1.26328, 1.41194]
        ])
        light_colors = [
            '#ffffff',
            '#ffffff',
            '#ffffff',
            '#ffffff',
            '#ffffff'
        ]
        z_dirs = np.arctan2(light_positions[:, 1], light_positions[:, 0])

        # Add lights to the scene
        for i in range(len(light_positions)):
            scene.spot_light(
                color=light_colors[i], intensity=intensity,
                angle=np.pi,
                ).rotate(0., 0., -z_dirs[i]).move(light_positions[i][0], light_positions[i][1], light_positions[i][2])


app.add_static_files('/geo', './')

camera=create_camera([0, -200., 1.25], np.pi / 6. )
with ui.scene(
	width=1024,
	height=800, 
	camera=camera, 
	grid=False, 
	) as ui_3d_scene:
        create_lights(ui_3d_scene, 60.)
        ui_3d_scene.stl(
                'geo/mean_all.stl' 
            ).rotate(np.pi / 2, 0., 0.).material(color='#000000')

        ui_3d_scene.gltf(
            f'geo/Configured_design_3D_sim.glb', 
        ).scale(0.01).rotate(np.pi / 2, 0., 0.).material(side='double')

ui.run()
```

**Comment by falkoschindler** [comment, marked as answer]
Thanks for providing the meshes, @maria-korosteleva! I think I understand what is happening.

Here is the essence of your current code:
```py
with ui.scene(camera=ui.scene.perspective_camera(fov=0.5)) as scene:
    scene.move_camera(y=-200, look_at_z=1)
    scene.spot_light(color='white', intensity=100, angle=3.14).move(y=-3, z=2)
    scene.stl('geo/mean_all.stl').rotate(1.57, 0, 0).material(color='black')
    scene.gltf('geo/Configured_design_3D_sim.glb').scale(0.01).rotate(1.57, 0, 0)
```
![Screenshot 2024-09-06 at 17 30 23](https://github.com/user-attachments/assets/6fce7184-3989-40f4-b735-59cf6167c5ef)

Your camera is 200 "meters" away from the object and the field of view is very narrow. I guess this causes resolution problems, because the camera has to render a very deep field of view. Changing the `near` and `far` parameters of the camera seems to solve the problem:
```py
with ui.scene(camera=ui.scene.perspective_camera(fov=0.5, near=190, far=210)) as scene:
    scene.move_camera(y=-200, look_at_z=1)
    ...
```
![Screenshot 2024-09-06 at 17 30 37](https://github.com/user-attachments/assets/40c8dd3d-ed01-42af-adb3-bca59c81f457)

Alternatively you can also use an orthographic camera, so that the y position becomes irrelevant:
```py
with ui.scene(camera=ui.scene.orthographic_camera(size=1.8)) as scene:
    scene.move_camera(y=-2, z=1, look_at_z=1)
    ...
```
![Screenshot 2024-09-06 at 17 33 39](https://github.com/user-attachments/assets/413daecf-13da-4cb9-8cd2-166635b44d81)

**Reply by maria-korosteleva** [reply]
Wow, @falkoschindler, that's amazing! Thank you so much for helping to figure this one out. 

After digging deeper into my code and experimenting with your suggestions, I discovered that the problem is even simpler. The perspective camera was not supposed to mimic the orthographic one. What happened is I copied FOV setup from another renderer I use, where radians are used as units (hence `np.pi/6`), while nicegui function uses degrees. So the intended 30 degrees turned into 0.5 degrees 😅 I didn't notice that, and instead adjusted the camera position to be further away to get some decent output, and then ran into rendering artifacts in the question😅

Once I adjusted the FOV to 30 degrees and put a closer camera location, the artifacts were gone as well, even with default near/far clipping plane values.  Thank you so much, I don't think I would have been able to catch that issue without your input 🙏
